File: cses/geometry/task2190-line-segment-intersection/main.py
# WA

import os
from sys import stdin
from typing import Callable, List, Tuple

dbg = bool(os.getenv("DEBUG"))
read_tokens: Callable[[], List[str]] = lambda: stdin.readline().split()


def det(p1: Tuple[int, int], p2: Tuple[int, int], p3: Tuple[int, int]) -> int:
    (x1, y1) = p1
    (x2, y2) = p2
    (x3, y3) = p3
    return (x2 - x1) * (y3 - y1) - (y2 - y1) * (x3 - x1)


def dot(p1: Tuple[int, int], p2: Tuple[int, int], p3: Tuple[int, int]) -> int:
    (x1, y1) = p1
    (x2, y2) = p2
    (x3, y3) = p3
    return (x2 - x1) * (x3 - x1) + (y2 - y1) * (y3 - y1)


def main_case() -> None:
    x1, y1, x2, y2, x3, y3, x4, y4 = map(int, read_tokens())
    p1 = (x1, y1)
    p2 = (x2, y2)
    p3 = (x3, y3)
    p4 = (x4, y4)

    # WC
    # det1 = det(p3, p4, p1)
    # det2 = det(p3, p4, p2)
    # det3 = det(p1, p2, p3)
    # det4 = det(p1, p2, p4)
    # if det1 * det2 <= 0 and det3 * det4 <= 0:
    #     print("YES")
    # else:
    #     print("NO")

    # WC
    # det3 = det(p1, p2, p3)
    # det4 = det(p1, p2, p4)
    # if not (det3 == 0 and det4 == 0):
    #     det1 = det(p3, p4, p1)
    #     det2 = det(p3, p4, p2)
    #     if det1 * det2 <= 0 and det3 * det4 <= 0:
    #         print("YES")
    #     else:
    #         print("NO")
    # else:
    #     dot2 = dot(p1, p2, p2)
    #     dot3 = dot(p1, p2, p3)
    #     dot4 = dot(p1, p2, p4)
    #     ok1 = det3 * det4 <= 0
    #     ok2 = 0 <= dot3 and dot3 <= dot2
    #     ok3 = 0 <= dot4 and dot4 <= dot2
    #     if ok1 and (ok2 or ok3):
    #         print("YES")
    #     else:
    #         print("NO")


def main() -> None:
    (t,) = map(int, read_tokens())
    for _ in range(t):
        main_case()


if __name__ == "__main__":
    main()

"""
python misc/run.py cses/geometry/task2190/main.py

%%%% begin
5
1 1 5 3 1 2 4 3
1 1 5 3 1 1 4 3
1 1 5 3 2 3 4 1
1 1 5 3 2 4 4 1
1 1 5 3 3 2 7 4
%%%%
NO
YES
YES
YES
YES
%%%% end

%%%% begin
1
-2 -5 7 -5 -3 -5 -4 -5
%%%%
NO
%%%% end
"""

File: cses/geometry/task2190-line-segment-intersection/main.rs
// AC

use std::{fmt::Debug, io, str::FromStr};

fn check_intersect(p: (i64, i64), l1: (i64, i64), l2: (i64, i64)) -> bool {
    if get_signed_area(p, l1, l2) != 0 {
        return false;
    }

    let x0 = l2.0 - l1.0;
    let y0 = l2.1 - l1.1;
    let x1 = p.0 - l1.0;
    let y1 = p.1 - l1.1;

    let dot = x0 * x1 + y0 * y1;
    let len0 = x0 * x0 + y0 * y0;
    let len1 = x1 * x1 + y1 * y1;

    dot >= 0 && len0 >= len1
}

fn get_signed_area(p: (i64, i64), q: (i64, i64), r: (i64, i64)) -> i64 {
    let x0 = q.0 - p.0;
    let y0 = q.1 - p.1;
    let x1 = r.0 - p.0;
    let y1 = r.1 - p.1;
    x0 * y1 - x1 * y0
}

fn main_case() -> io::Result<()> {
    let v = read_tokens::<i64>()?; // [-10^9, 10^9];
    let p1 = (v[0], v[1]);
    let p2 = (v[2], v[3]);
    let q1 = (v[4], v[5]);
    let q2 = (v[6], v[7]);

    // check singular case (one endpoint is on the other segment)e
    if check_intersect(p1, q1, q2)
        || check_intersect(p2, q1, q2)
        || check_intersect(q1, p1, p2)
        || check_intersect(q2, p1, p2)
    {
        println!("YES");
        return Ok(());
    }

    // check non-singular case
    let p1_p2_q1 = get_signed_area(p1, p2, q1).signum();
    let p1_p2_q2 = get_signed_area(p1, p2, q2).signum();
    let q1_q2_p1 = get_signed_area(q1, q2, p1).signum();
    let q1_q2_p2 = get_signed_area(q1, q2, p2).signum();
    if p1_p2_q1 * p1_p2_q2 == -1 && q1_q2_p1 * q1_q2_p2 == -1 {
        println!("YES");
        return Ok(());
    }

    println!("NO");
    Ok(())
}

fn main() -> io::Result<()> {
    let t: usize = read_tokens()?[0]; // [1, 10^5]
    for _ in 0..t {
        main_case()?
    }
    Ok(())
}

/*
python misc/run.py cses/geometry/task2190-line-segment-intersection/main.rs

%%%% begin
5
1 1 5 3 1 2 4 3
1 1 5 3 1 1 4 3
1 1 5 3 2 3 4 1
1 1 5 3 2 4 4 1
1 1 5 3 3 2 7 4
%%%%
NO
YES
YES
YES
YES
%%%% end
*/

//
// utils
//

fn read_tokens<F: FromStr>() -> io::Result<Vec<F>>
where
    <F as FromStr>::Err: Debug,
{
    let mut line = String::new();
    io::stdin().read_line(&mut line)?;
    let mut result: Vec<F> = Vec::new();
    for token in line.trim().split(" ") {
        let value: F = token
            .parse()
            .map_err(|e| io::Error::new(io::ErrorKind::Other, format!("{:?}", e)))?;
        result.push(value);
    }
    Ok(result)
}
